wavesexchange_repos/src/circuit_breaker/mod.rs

```diff
@@ -0,0 +1,299 @@
+pub mod config;
+
+pub use config::Config;
+use wavesexchange_log::debug;
+
+use std::{
+    future::Future,
+    mem::drop,
+    sync::Arc,
+    time::{Duration, Instant},
+};
+use tokio::sync::RwLock;
+
+/// A trait represents some data source that can fail.
+/// Must be implemented for a struct that used in circuit breaker.
+pub trait FallibleDataSource {
+    type Error;
+
+    /// Shows if error will increase CB's counter
+    fn is_countable_err(err: &Self::Error) -> bool;
+
+    /// Set up CB's behaviour after maximum errors limit exceeded
+    fn fallback(&self, elapsed_ms: u128, err_count: usize) -> Self::Error {
+        panic!(
+            "CircuitBreaker panicked after {err_count} errors in a row happened in {elapsed_ms}ms"
+        )
+    }
+}
+
+pub trait DataSrcInitFn<S: FallibleDataSource>:
+    Fn() -> Result<S, S::Error> + Send + Sync + 'static
+{
+}
+
+impl<T, S: FallibleDataSource> DataSrcInitFn<S> for T where
+    T: Fn() -> Result<S, S::Error> + Send + Sync + 'static
+{
+}
+
+/// Count erroneous attempts while quering some data source and perform reinitialization/fallback.
+///
+/// To use within an object, you must implement `FallibleDataSource` first.
+///
+/// Example:
+/// ```rust
+/// fn main() {
+///     struct Repo;
+///     struct RepoError;
+///
+///     impl FallibleDataSource for Repo {
+///         type Error = RepoError;
+///
+///         fn is_countable_err(err: &Self::Error) -> bool {
+///             true
+///         }
+///     }
```

Comment:
What is the use of implementing some trait for that Repo, if methods of that trait never receive instance of that Repo? Why can't it be a separate entity then? Looks like unnecessary complication.

Reply:
For example, to implement this (foreign) trait on a connection pool (which is also a foreign type) one needs to wrap that connection pool to a neotype due to the Orphan Rule, which seems unnecessarily complicated.